skill-tree/src/main/java/com/RDS/skilltree/Filters/JWTAuthenticationFilter.java

```diff
@@ -0,0 +1,64 @@
+package com.RDS.skilltree.Filters;
+
+import com.RDS.skilltree.Authentication.UserAuthenticationToken;
+import com.RDS.skilltree.utils.JWTUtils;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.Cookie;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+import org.springframework.web.util.WebUtils;
+
+import java.io.IOException;
+
+@Slf4j
+public class JWTAuthenticationFilter extends OncePerRequestFilter {
+
+    @Value("${cookieName}")
+    private String cookieName;
+    @Autowired
+    private JWTUtils jwtUtils;
+
+    @Override
+    public void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+
+        String token = getJWTFromRequest(request);
+
+
+            if (StringUtils.hasText(token) && jwtUtils.validateToken(token)) {
+                String rdsUserId = jwtUtils.getRDSUserId(token);
+                String role = jwtUtils.getUserRole(token);
+
+
+                UserAuthenticationToken authentication = new UserAuthenticationToken(role, rdsUserId);
+                authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
+                SecurityContextHolder.getContext().setAuthentication(authentication);
+            }
+
+
+
+        filterChain.doFilter(request, response);
+    }
+
+    public String getJWTFromRequest(HttpServletRequest request) {
+
+        /* check for cookie */
+        Cookie RDScookie = WebUtils.getCookie(request, cookieName);
+        if (RDScookie != null) return RDScookie.getValue();
+
+        /* extract token from header */
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
```

Comment:
Avoid using magic numbers! I know logically you are reading from the 7th position, but it's better to not have a random number suddenly appearing.

Someone new (with no context) reading the codebase would think how did you come up with 7?

Try using "Bearer ".length().
Also, maintain the strings "Authorization" and "Bearer " as `private static final` class variables.

Also,

Reply:
done, thanks for this suggestion